clang/lib/Driver/Driver.cpp

```diff
@@ -4122,7 +4122,11 @@ class OffloadingActionBuilder final {
             // header.
             Action *CompileAction =
                 C.MakeAction<CompileJobAction>(A, types::TY_Nothing);
-            DA.add(*CompileAction, *ToolChains.front(), nullptr,
+            // If the bound arch is provided make sure to propagate it.
+            const char *BoundArch = !SYCLTargetInfoList.empty()
+                                        ? SYCLTargetInfoList.back().BoundArch
+                                        : nullptr;
+            DA.add(*CompileAction, *ToolChains.front(), BoundArch,
```

Comment:
you have `back` then `front`, does that works on multi targets ?

Probably best to iterate over a zip view of `SYCLDeviceActions` and `SYCLTargetInfoList` (there is one action entry per target info) as done in some functions below.

so,
```
          for (Action *&A : SYCLDeviceActions) {
```

can become

```
      for (auto TargetActionInfo :
           llvm::zip(SYCLDeviceActions, SYCLTargetInfoList)) {
        Action *A = std::get<0>(TargetActionInfo);
        DeviceTargetInfo &TargetInfo = std::get<1>(TargetActionInfo);
```

and then

```suggestion
            DA.add(*CompileAction, *TargetInfo.TC, TargetInfo.BoundArch,
```

This will ensure consistency between toolchain and boundarch.

Reply:
I really like the idea of zipping those two up, added.